056-zk-kraft-migration.md

```diff
@@ -0,0 +1,258 @@
+So the prerequisite is to have the Kafka cluster still using ZooKeeper but running the Kafka brokers with `KafkaNodePool`(s).
+The user can follow the procedure described [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) in the official documentation, in order to meet this prerequisite.
+
+Before migrating to KRaft, the cluster must be upgraded to Kafka version minimum of 3.5.0 with `inter.broker.protocol.version` set to “3.5”.
+
+Furthermore, the running Apache Kafka cluster should not be using JBOD because in this case the migration to KRaft is not supported.
+
+### Migration
+
+The migration process is semi-automatic which means having the user driving the process by interacting with the operator (mostly via annotations and custom resources creation/deletion) which is in charge to do the rest (i.e. nodes re-configuration, rolling, checking migration status, ...).
+
+> NOTE: this kind of approach is already in place for the Kafka version upgrade.
+
+Assume the following prerequisites, with a ZooKeeper based Kafka cluster defined by:
+
+* A `Kafka` custom resource.
+* One or more `KafkaNodePool`(s) for running the ZooKeeper based brokers.
+
+If the operator is not using KRaft by default nor has the `+UseKRaft` feature gate enabled, first step for the user would be enabling the `+UseKRaft` feature gate unless it’s already enabled by default in a new operator release (depending on which Strimzi version this proposal will land).
+So the current operator may need to be upgraded to the new release.
+In both cases, the operator is rolled, a new one starts and it’s able to handle the creation of new Apache Kafka clusters in KRaft mode but not migrating the old ZooKeeper based ones automatically.
+
+Currently, the operator prints the following warning:
+
+```shell
+io.strimzi.operator.common.model.InvalidResourceException: Tha Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
+```
+
+This is because with KRaft mode enabled in the operator, the reconciled `Kafka` resource doesn’t have any controller nodes (in any pool).
+
+The migration process can be also rolled back to ZooKeeper at any point before taking the KRaft controllers out of the migration mode.
+
+### Finite State Machine (FSM)
+
+The migration process can be described with a Finite State Machine (FSM).
+The operator has a new `KRaftMigrationManager` component in charge of driving and tracking transitions in the FSM.
+Transitioning across the states depends on the current state, stored in the `strimzi.io/kraft-migration-state` annotation on the `Kafka` custom resource, and the "input" represented by the user interaction with the operator through the `strimzi.io/kraft-migration` annotation to be applied on the `Kafka` custom resource.
+
```

Comment:
1. As I said before, I think you are making the mistake of building this on the current `UseKRaft` feature gate. That is currently written for development and testing only. For taking it further towards production, it will have to evolve and likely have its own annotations. So you are either missing those annotations - or maybe better, you should define and reuse them. I think this heavily impacts anything from this line on.
2. Storing state in the annotation of Kafka CR is not possible. You risk other tooling to take it away from you. You cannot use user managed custom resources to store any annotations.

Reply:
Sorry but I haven't got what you mean by point 1.

> As I said before, I think you are making the mistake of building this on the current UseKRaft feature gate. That is currently written for development and testing only. For taking it further towards production, it will have to evolve and likely have its own annotations.

What does it mean? We should stop and wait for when Strimzi has a full KRaft support by waiting for when Kafka has the full KRaft support? 

> So you are either missing those annotations - or maybe better, you should define and reuse them. I think this heavily impacts anything from this line on.

There are no annotations related to KRaft AFAIK right now in Strimzi so I am defining the ones I need for this implementation. What's wrong with that?

> Storing state in the annotation of Kafka CR is not possible. You risk other tooling to take it away from you. You cannot use user managed custom resources to store any annotations.

The initial proposal was based on using a field in the `Kafka.status` and you raised the good point that it's a change in the API for a temporary thing so I agreed. The idea I had was using an annotation. I can't think about another way (a totally external resources? i.e. a ConfigMap). What are you proposing? We need to know the state of the migration and persisting it somewhere.

Reply:
No, I'm not necessarily suggesting you should wait, although that would give us much more clarity obviously. But I do not think there is anything wrong with planning for the future. 

However, I think planning it upfront requires you to predict or design how the future will look like. For example, if you base this on the current state, then you would migrate all clusters managed by a single operator in parallel. I don't think that would work. So you need to figure out how it will actually look like at the time when the migration makes sense. Because I think writing the proposal now based on the current state and rewriting it later because things changed makes little sense.

>The initial proposal was based on using a field in the Kafka.status and you raised the good point that it's a change in the API for a temporary thing so I agreed. The idea I had was using an annotation. I can't think about another way (a totally external resources? i.e. a ConfigMap). What are you proposing? We need to know the state of the migration and persisting it somewhere.

Yeah, well. What should I tell you ... not every reviewer gives you good comments 🤷‍♂️. But I think the comment about the annotation stands -> if you wanna use annotation, you have to place it on resources we manage, not the user as you will easily lose it there. I guess the comment about the status fields stands, but maybe we can live with it as we likely drop it in `v1` APi? I guess it also relates to the comments about how users monitor the steps. For internal things, maybe you can use annotations on something internal. But if users is expected to watch the state, maybe status is the best despite the negative aspects.

Reply:
> No, I'm not necessarily suggesting you should wait, although that would give us much more clarity obviously. But I do not think there is anything wrong with planning for the future.

However, I think planning it upfront requires you to predict or design how the future will look like. For example, if you base this on the current state, then you would migrate all clusters managed by a single operator in parallel. I don't think that would work. So you need to figure out how it will actually look like at the time when the migration makes sense. Because I think writing the proposal now based on the current state and rewriting it later because things changed makes little sense.

I am confused. This sentences seem to say something and the opposite :-) So I am not sure what should be our plan for migration tbh.

> I guess the comment about the status fields stands, but maybe we can live with it as we likely drop it in v1 APi?

Well I read your previous comment as totally against to use a field in the `status` then I suggested the annotation. These are the only two options we usually have for this kind of things.

> I guess it also relates to the comments about how users monitor the steps. For internal things, maybe you can use annotations on something internal. But if users is expected to watch the state, maybe status is the best despite the negative aspects.

Yeah I left a comment there about checking the annotation based on your negative comment on the `status` field.
Of course, watching an annotation is not that great.

> What should I tell you ... not every reviewer gives you good comments 🤷‍♂️

I think that sometime much clarity is better and proposing an idea can help when commenting.

Reply:
> Well I read your previous comment as totally against to use a field in the status then I suggested the annotation. These are the only two options we usually have for this kind of things.

Well, sometimes, things are not great but they are still the best option :-/

> I am confused. This sentences seem to say something and the opposite :-) So I am not sure what should be our plan for migration tbh.

From my point of view, I basically think you have two choices -> wait with the proposal until it is clear ... or come up with how it will actually work when it matters. The second option might be best as you would need to control when does what part of the code base take over.

I think the general idea is simple. Latest once the `UseKRaft` FG is ready to move beyond development and testing only (which I guess at minimum needs the rolling updates and liveness / readiness, possibly upgrades), we would need to have some annotation similar to the node pools one to be used when the FG is enabled but only some clusters should use KRaft. And that will change this proposal. I think there is an opportunity to use the same annotation to drive the migration. But if nothing else, the migration process would need adding this annotation at some stage.

Reply:
> Well, sometimes, things are not great but they are still the best option :-/

I will revert back the usage of the field then which was anyway my preferred option.

> I think the general idea is simple. Latest once the UseKRaft FG is ready to move beyond development and testing only (which I guess at minimum needs the rolling updates and liveness / readiness, possibly upgrades), we would need to have some annotation similar to the node pools one to be used when the FG is enabled but only some clusters should use KRaft. And that will change this proposal. I think there is an opportunity to use the same annotation to drive the migration. But if nothing else, the migration process would need adding this annotation at some stage.

Of course, the migration process needs rolling controllers but it's something that Kate & Tina are already addressing. Something working is already available in their fork so I can anyway start exploring migration using that rolling feature, hoping that their work will be soon in main branch.
Regarding annotations, I guess you are referring to the `strimzi.io/node-pools` one. We don't have KRaft annotations right now and I don't see why I should wait for other annotations to be in place if I can have a dedicated one `strimzi.io/kraft-migration` used for migration purposes. It's just an annotation and when migration won't be useful anymore, the annotation won't be handled anymore. I don't see it as a big problem tbh. It's much better instead of waiting for some other KRaft work being done, adding annotations and then using one of them for migration. It would be just a blocker for this work.

Tbh what I see really a blocker right now is the need for rolling controllers (which is available in some shape in the Kate & Tina fork). The rest of KRaft missing features is not a blocker for the migration itself. They are a blockers for KRaft in GA support from Strimzi, I would agree on this. But waiting for a KRaft ready both in Kafka and Strimzi before starting a migration work would mean postponing this work for months and months.